File: challenges/2026-09-21-harmonic-pair-counter/solution.py
from typing import List


def count_harmonic_pairs(freq: List[int], k: int) -> int:
    """Return the number of index pairs (i, j), i < j, whose gcd(freq[i],
    freq[j]) is exactly k."""
    # TODO: implement
    pass


def _run_self_checks() -> None:
    cases = [
        ("example 1", [4, 6, 8, 10, 3], 2, 5),
        ("example 2", [2, 4, 8, 16], 4, 2),
        ("example 3", [5, 7, 11], 5, 0),
        ("edge: k larger than every value", [2, 4, 6], 100, 0),
        ("edge: all elements equal to k", [7, 7, 7], 7, 3),
    ]

    for name, freq, k, expected in cases:
        actual = count_harmonic_pairs(freq, k)
        status = "PASS" if actual == expected else "FAIL"
        print(f"[{status}] {name}: expected={expected!r} actual={actual!r}")


if __name__ == "__main__":
    _run_self_checks()

File: challenges/2026-09-21-harmonic-pair-counter/solution.ts
function countHarmonicPairs(freq: number[], k: number): number {
  // TODO: implement
  return 0;
}

function runSelfChecks(): void {
  const cases: [string, number[], number, number][] = [
    ["example 1", [4, 6, 8, 10, 3], 2, 5],
    ["example 2", [2, 4, 8, 16], 4, 2],
    ["example 3", [5, 7, 11], 5, 0],
    ["edge: k larger than every value", [2, 4, 6], 100, 0],
    ["edge: all elements equal to k", [7, 7, 7], 7, 3],
  ];

  for (const [name, freq, k, expected] of cases) {
    const actual = countHarmonicPairs(freq, k);
    const status = actual === expected ? "PASS" : "FAIL";
    console.log(`[${status}] ${name}: expected=${expected} actual=${actual}`);
  }
}

runSelfChecks();
